### ISSUE TITLE
Crippled Lapack Declaration on Windows

### ISSUE THREAD

**Issue by kthohr** [RcppCore/RcppArmadillo#207]
Greetings,

A lot of Lapack routines were added with R version 3.3.0 (zgges, etc.), and the package configure script for Unix-alike systems correctly identifies this. However, for Windows systems, the assumption in RcppArmadilloConfig.h is that these routines are still missing. With R v3.5 being released in the next month or so, is now a good time to (safely) change this default behavior and assume most users have updated?

Best,
Keith

**Comment by eddelbuettel** [comment]
No, there isn't, sadly, as one @baptiste 's package still failed over a complex decomposition.  As long as the (partial) LAPACK copy that ships with R and is used by default on Windows, we are bound to have these issues.  I fought this twice already.

**Comment by eddelbuettel** [comment]
What is your underlying question?  Are you missing a particular routine?  Over the years it was generally just the somewhat more esoteric complex functions that were seen missing, and R Core helped twice and added them.

In general, until the R build changes and always uses an external (complete) LAPACK I fear that is little we can do better.  Whenever we have left the guard down and said "we're good" we ended up being bitten by something we actually did not have.

**Comment by eddelbuettel** [comment]
It think when @baptiste was bitten last it was due to a change on the Armadillo side when @conradsnicta  invoked a different (sparse?) solver.  I am a little fuzzy on the details of that conversation -- may have been mostly on the r-package-devel or rcpp-devel mailing lists.

As for 
> there are no issues

that tends to be my sentiment too.  As of today, we do have 458 packages on CRAN that use RcppArmadillo.  In general, things just work. 

This [README](https://github.com/wch/r-source/blob/trunk/src/modules/lapack/README) in the R sources has a good list of what is included.  Again, for RcppArmadillo, _ex ante_ we do not know which library we are facing hence the need to dial down.  So we set `ARMA_CRIPPLED_LAPACK` if we're on Windows, or if we see `Rlapack.*` being used.  That `#define` then affects Conrad's code.

**Comment by kthohr** [comment]
Sorry I wasn't clear with that example. Right now qz() for complex-valued matrices is ifdef'd out on the Armadillo side, as zgges was missing until R version 3.3.0. However, I've tested RcppArmadillo on Windows with that guard removed and it works fine. So my feeling is that it should be okay to start removing _some_ of the guards based on the 3.3 update.

**Comment by eddelbuettel** [comment]
And again, to be crystal clean, that guard is in the Armadillo code, correct?

We could test this in a local patch we could try to noodle through CRAN.  If it all works we can discuss with Conrad if/where/how the #ifdef is still needed.  A quick recursive grep reveals that it is only used in 
`inst/include/armadillo_bits/auxlib_meat.hpp` but I fear we have to test each and every use case against different R versions and their `Rlapack.*`.  (Or at least 3.3.0 and later...)

That is quite a bit of work.  Are you volunteering?

**Comment by kthohr** [comment]
Yep, that guard is on the Armadillo side. 

I'm happy to do the heavy lifting on this---I'm familiar with Rlapack and the internal Armadillo code. Would you prefer to start with the single qz case and then branch out to other functionality?

**Comment by kthohr** [comment]
Sounds good to me.

**Comment by baptiste** [comment]
For the record since my cda package was mentioned: the package actually broke for Linux on CRAN, which apparently also rely on the internal R Lapack (I thought it did only for Windows). The package itself does not use any unusual routine, it just calls solve() on a standard complex linear system, but Armadillo recently introduced a new special-case for banded matrices that would require, if it was ever called, a Lapack routine that R doesn't provide. Linux CRAN checks find this routine in the headers missing, and break (in contrast Windows is always flagged as crippled). Someone kindly provided a workaround (to basically flag Lapack as crippled) on the mailing list so that it would pass CRAN checks like on Windows, but I never found the time to test this (when I tried on my Linux box an external Lapack was used, and when I tried to switch that didn't seem to affect anything; maybe I'd need to compile R with some flags for the purpose of this test).

Bottom line: cda is now archived on CRAN; I hope to return to it someday. 

It sounds like we're facing opposing problems: you want to tell CRAN not to flag 'crippled' how that a routine is available, while I'd need to enforce the flag manually because some routine is missing.

**Comment by eddelbuettel** [comment]
> I thought it did only for Windows

It's a compile time option. On the non-Windows systems, if you do `configure --help` you get (inter alia)

```
  --with-blas             use system BLAS library (if available), or specify it [no]
  --with-lapack         use system LAPACK library (if available), or specify it [no]
```

so the default may actually be to use the 'crippled' lapack (I think my local r-devel builds do that as I don't override it).  But most distribution builds will probably use the external BLAS.

I supposed one can alter it on Windows too which must be how Microsoft injects the MKL.

Bottom line: "it's complicated" and when we build RcppArmadillo we have to track it.

**Comment by conradsnicta** [comment]
I really don't understand why R simply doesn't ship with a complete copy of LAPACK. It's actually more work for them to keep their own crippled, borked and incomplete version. Is this due to some antiquated notion of saving storage space? Or maybe they're just closet fans of S&M?

As of Armadillo 8.300, the following "extra" lapack routines are used:
spbtrf, dpbtrf, cpbtrf, zpbtrf
sgbsv, dgbsv, cgbsv, zgbsv
sgbsvx, dgbsvx, cgbsvx, zgbsvx

These functions have been in LAPACK since pretty much "forever". Moreover, they deal with systems that are often encountered in real-life applications (including financial data analysis).

**Comment by baptiste** [comment]
@conradsnicta I think a suitably edited version of your question would be a useful discussion to have on R-devel. I don't know the reasons myself, but my guess would be that R has always been very conservative with respect to numerical accuracy, and R-core may have found it necessary to manually patch/review some routines on a case-by-case basis. Or it could be that the effort of writing all the appropriate wrappers and ensuring their correctness has always been the limiting factor, leading to this cherry-picked, add-only-when-necessary situation.

**Comment by conradsnicta** [comment]
@baptiste - If R-core doesn't trust lapack, then they should either (i) submit their patches upstream, or (ii) get out of the business of providing numerical software. Honestly, there is no excuse in shipping a borked lapack and then wondering why people complain. I'm sure there was a good reason ~200 million years ago (when dinosaurs roamed the earth) for cherry picking lapack routines, but these days the world is a different place.

**Comment by baptiste** [comment]
@conradsnicta as I said I don't know the reasons (and therefore should not have replied); I made a wild guess based on my recollection of  discussions around flavours of Blas and Lapack (cf openBLAS and other alternatives), with R favouring a conservative (inefficient but more reliably accurate) choice by default. It may very well be completely irrelevant here; R-devel would have much more pertinent things to say.

**Comment by eddelbuettel** [comment]
This is a 30 year old discussion.  Statistician prefer eg the QR solver with pivoting which is why a modified LINPACK gets used in the standard `lm()`.  But these days with double precision you have to work hard for finding actual differences.  The example `fastLm()` in RcppEigen has a worked example thanks to @dmbates , the one in for `fastLm()` in RcppArmadillo does not but at least I put [a note in the documentation](https://github.com/RcppCore/RcppArmadillo/blob/master/man/fastLm.Rd#L26-L54).  

And for 99.95% of all users the shipped LAPACK subset is actually sufficient.  In all those years the only user I encountered who had issues was @baptiste .  Now, his uses were all legit so we tried (and managed) to get this improved -- which can be done.  But running around, screaming murder and calling other people names rarely does.

**Comment by conradsnicta** [comment]
@eddelbuettel - The LINPACK vs LAPACK angle is a red herring in my view. While the two libraries share some heritage, they are two separate and distinct beasts.

Properly linking with one library does not preclude linking with the other library, unless there is a name clash in a flat namespace (from the linker's point of view). On operating systems from within the last decade that is easily fixable (eg. through a separate wrapper), without resorting to cherry-picking lapack functions.

In other words, so far I see no valid reasons for shipping a borked lapack. I do understand that there is a high level of technical debt within the R code base, but piling more hacks on top of it is not the sustainable way to solve it. There is a clear and sustainable solution: ship complete lapack. If R needs to also link linpack, nobody is stopping that.

**Comment by eddelbuettel** [comment]
The part I have control over -- how R is built and shipped for Debian and hence Debian derivatives -- does that: full external Lapack.  Other parts and builds I cannot control, but I invite you to engage with R Core and argue for change there.

**Comment by baptiste** [comment]
I've emailed R-devel.

**Comment by eddelbuettel** [comment]
Thanks.  In the meantime we can try refining the tests.

**Comment by kthohr** [comment]
Thanks for starting the discussion, @baptiste . This is a bane of mine, too.

On the `CRIPPLED_LAPACK` side, I've identified which guards can be removed; I'm just figuring out an intelligent way to test this to unit.tests. Dirk, would you prefer I add a new set of tests (say runit.Rlapack)?

**Comment by eddelbuettel** [comment]
Awesome news @kthohr ! 

Maybe we can discuss a bit more.  What we currently have is blunt: one test, one variable, being used in a good half dozen instances inside that 'meat' file.   Do you think we should have variables, or just do away with the ones you identified as being available?

As for new tests, if it is a new topic, a new top-level R file `inst/unitTests/runit.$name.R` with a new matching `cpp/$name.cpp` seems the best.  `Rlapack` works for me as `$name`.

**Comment by eddelbuettel** [comment]
Oh, and should we _remove_ the guard or _refine the test_ to use when we know it works?  Where do you think is the cutoff?  R 3.3.0* ? That would be old enough not to worry.

**Comment by kthohr** [comment]
I think we should remove the guards entirely, where possible. Here's a list of affected functions (with cutoffs):
```
eig_pair for complex matrices
Lapack: zggev.  Added: R 3.3.0

qz complex matrices
Lapack: zgges.  Added: R 3.3.0

svd_dc and svd_dc_econ for complex matrices
Lapack: zgesdd. Added: R 3.1.0

solve_square_refine for complex matrices
Lapack: zgesvx. Added: R 3.3.0

solve_approx_svd for complex matrices
Lapack: zgelsd. Added: R 3.1.0

schur for complex matrices
Lapack: zgees.  Added: R 3.3.0
```
So some are related to changes made in R version 3.1.0, but most are from version 3.3.0. The general `CRIPPLED_LAPACK` guard will remain for some decompositions of band matrices (see `chol_band` and `solve_band_fast`).

**Comment by kthohr** [comment]
So my quick and dirty changes can be found here:
```R
install_github("kthohr/RcppArmadillo",ref="lapack-checks")
```
(Tests are in RcppArmadillo.cpp instead of a separate unit test file, while I will of course change for the PR.) I have checked this on a Windows machine with our beloved Rlapack. `ARMA_EXTRA_DEBUG` is enabled, so calling
```R
library(RcppArmadillo)

n <- 5

norm(cx_eig_pair_test(n),"2")
norm(cx_qz_test(n),"2")
cx_rank_test(n) - n
norm(cx_solve_test(n),"2")
norm(cx_pinv_test(n),"2") - 1
norm(cx_schur_test(n),"2")
```
will show the internal pings of the Lapack functions. Any volunteers with broken Lapack distributions: please check this if you can.

Would the following make most sense?
* Create a pull request with @conradsnicta to remove the relevant guards on the Armadillo side.
* Then a pull request to RcppArmadillo with the `runit.Rlapack` tests.

**Comment by eddelbuettel** [comment]
Good plan. I agree that we should just remove the relevant test on the side of @conradsnicta / Arma.

I think I will play with your snippet above and submit this to the various winbuilders and friends.  Maybe also add a simple 

```c++
// [Rcpp::export]]
int whatAreWe() {
#if defined(ARMA_CRIPPLED_LAPACK)
   return 1;
#else 
   return 0;
#endif;
}
```

to know where we are at.

**Comment by kthohr** [comment]
Great! I just added a `armadillo_Rlapack()` function to cover your suggestion.

On my end, I tested the code with fresh installations of R---versions 3.3.0 and 3.4.3---on Windows. I also checked against R-devel on my main system (macOS), but that uses a system supplied Lapack (the Accelerate library).

**Comment by eddelbuettel** [comment]
Are you set up for R-hub? That should give you a wide variety of different systems?  (And I am tied up with something else...)

**Comment by kthohr** [comment]
Nope, never worked with R-hub. And no rush on my side.

**Comment by eddelbuettel** [comment]
It is _the awesomest_ thing ever.  Brew a fresh cup and look into it:

![image](https://user-images.githubusercontent.com/673121/37939451-75e10c68-3128-11e8-81af-6f5dd56f4aad.png)

That is the webview at https://builder.r-hub.io/; I tend to use rhub package...

**Comment by conradsnicta** [comment]
I'd suggest to first have all the changes present in an RcppArmadillo release candidate submitted to CRAN. Testing done by CRAN can turn up issues that we haven't thought of (corner cases, weirdo architectures, etc).  If all is okay, the upstream sources in Armadillo can then be modified.

I'd also suggest to do a paranoid check of all the lapack and blas symbols used by Armadillo against the symbols present in the borked R versions of lapack and blas.

https://github.com/conradsnicta/armadillo-code/blob/unstable/include/armadillo_bits/def_lapack.hpp
https://github.com/conradsnicta/armadillo-code/blob/unstable/include/armadillo_bits/def_blas.hpp

Caveat: don't rely on the above two lists to be static; as Armadillo gets new features, the above lists are likely to expand.

**Comment by coatless** [comment]
@conradsnicta to get a fixed link (e.g. current commit / the [canonical form](https://help.github.com/articles/getting-permanent-links-to-files/)) press "Y" and then copy the URL:

https://github.com/conradsnicta/armadillo-code/blob/f979b6c8e6bbe7ff9457fd39d2f01e729919085f/include/armadillo_bits/def_lapack.hpp
 
https://github.com/conradsnicta/armadillo-code/blob/f979b6c8e6bbe7ff9457fd39d2f01e729919085f/include/armadillo_bits/def_blas.hpp

All keyboard shortcuts c.f.

https://help.github.com/articles/using-keyboard-shortcuts/#source-code-browsing

**Comment by eddelbuettel** [comment]
It's a little different.  Release candidates do not go to CRAN. I only upload to CRAN once every two months -- it is too involved otherwise. Details in the [ChangeLog](https://cloud.r-project.org/web/packages/RcppArmadillo/ChangeLog)  and [NEWS](https://cloud.r-project.org/web/packages/RcppArmadillo/news.html) at the package page.

I do roll up interim releases, and do eg the reverse-depends checking I did today.  But the systems used for those have full LAPACK anyway.

So I would suggest the following.  We really only need two blocks commented out, so I suggest that for the upcoming release we keep things mostly as they are but change 
 
```c++
#if defined(ARMA_CRIPPLED_LAPACK)    
```

to 

```c++
#if defined(ARMA_CRIPPLED_LAPACK) || defined(ARMA_CRIPPLED_LAPACK_BANDED)   
```

and I define that new variable, but not the old one.  That gives us the 2 out of 7 we need.   If that passes, we clean up for the next release on remove the old define.

Thoughts?

**Comment by kthohr** [comment]
@eddelbuettel - I don't think I follow the logic of that... where are we placing `ARMA_CRIPPLED_LAPACK_BANDED`?

For reference, this is what I was going to submit on the Armadillo side:
https://github.com/kthohr/armadillo-code/commit/1ac3e5de5c41fd52aabf5fa5e1a36c40fc08cf3a

Are we still keeping the guards around those blocks? Sorry for my confusion...

**Comment by eddelbuettel** [comment]
What I had in mind would simply be less invasive as a start.  We simply add a new "or" with the new define in the two spots where banded is used.

**Comment by conradsnicta** [comment]
I'm not keen on adding another define to deal with borked lapack.  In effect we would be then dealing with 3 versions of lapack: (i) proper lapack, (ii) very borked lapack, (iii) less borked lapack. That's creating too many test scenarios and adds to the maintenance burden.

The ideal solution is to support only (i), ie. talk some sense into R developers. If that's not doable, then let's aim to support (i) and (iii), by ensuring that a minimum allowed version of R is used (ie. the minimum version that has the less borked lapack on windoze). If a user has an ancient version of R with the super borked lapack, they're out of luck.

**Comment by coatless** [comment]
In essence, this means that _RcppArmadillo_ will need to depend on an _R_ version greater than or equal to R 3.3.0 (per @kthohr's analysis). If the petition to r-devel is successful for R 3.5.*, then two years from now the dependency could increase to R 3.5.0 (unlikely due to the release schedule).

**Comment by conradsnicta** [comment]
According to https://cran.r-project.org/bin/windows/base/old/ R 3.3 for windoze was released in April 2016, so about 2 years ago. Enough time has passed to ensure wide distribution.

**Comment by eddelbuettel** [comment]
1. R 3.5.0 will come out next months, nothing of this sort will go into it.
2. @kthohr did a fine analysis, we can easily depend on R (>= 3.3.0) as we have similar / worse depends on a new enough compiler 
3. It is *not* just Windows.  It is every possible R build where external lapack is not switched on.
4. It is in fact '(i) proper lapack, (iii) less borked lapack' as we are down to two instances of banded matrices.

**Comment by eddelbuettel** [comment]
So I took the version of [`auxlib_meat.hpp`](https://github.com/kthohr/armadillo-code/commit/1ac3e5de5c41fd52aabf5fa5e1a36c40fc08cf3a) from @kthor and created a new minor test release in a branch.

That passes on the four default [R Hub](https://builder.r-hub.io/) instances invoked from `rhub::check_for_cran()`.    In that version, we are down to four uses of the `#define`.  I think this could go into Armadillo 8.500 as is.

**Comment by eddelbuettel** [comment]
And if anyone wants to play along, this is now in [this RcppArmadillo branch](https://github.com/RcppCore/RcppArmadillo/tree/feature/less_crippled_lapack).  Might even fold that into master.  

(@kthohr I used your school email for the ChangeLog as you list it on you GH profile.  We can adjust if you'd prefer a different one.)

**Comment by kthohr** [comment]
Thanks, Dirk! I created a PR to push a unit test into that branch.

(I've also changed my e-mail on my GH profile.)

**Comment by eddelbuettel** [comment]
This is now merged in the master branch (as well as upstream for the actual change), so we could close this ticket -- or keep it open as a reminder of the four remaining `#if defined` tests.

Thoughts?

**Comment by kthohr** [comment]
Great! I think we can close the ticket now.

**Comment by eddelbuettel** [comment]
I think you are right -- this was after all about reflecting what R 3.3.0 brought us -- so thanks again for all your work on this.

**Comment by coatless** [comment]
Need to add an _R_ depends on _R_ 3.3.0

MIA from: 

https://github.com/RcppCore/RcppArmadillo/pull/211
https://github.com/RcppCore/RcppArmadillo/pull/212

**Comment by eddelbuettel** [comment]
Sure, but we haven't released to CRAN either.  On the TODO list.
